### PR TITLE
[CALCITE-2818] EXTRACT returns wrong results for DATE and TIMESTAMP values before epoch

### PR DESCRIPTION
**Pull request by hsyuan**
JIRA: https://issues.apache.org/jira/browse/CALCITE-2818

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 1]
```diff
@@ -6713,6 +6713,57 @@ private void checkNullOperand(SqlTester tester, String op) {
...
+        "extract(minute from TIMESTAMP '1969-12-31 21:13:17.357')",
+        "13",
+        "BIGINT NOT NULL");
+    tester.checkScalar(
+        "extract(second from TIMESTAMP '1969-12-31 21:13:17.357')",
+        "17",
+        "BIGINT NOT NULL");
+    tester.checkScalar(
+        "extract(millisecond from TIMESTAMP '1969-12-31 21:13:17.357')",
+        "17357",
```
Comment (asereda-gs) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Why millisecond from `21:13:17.357` is `17357 ` and not `357` ?

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 2]
```diff
@@ -6713,6 +6713,57 @@ private void checkNullOperand(SqlTester tester, String op) {
         "extract(microsecond from cast(null as time))");
   }
 
+  @Test public void testExtractWithDatesBeforeUnixEpoch() {
+    tester.checkScalar(
```
Comment (asereda-gs) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Can you pls also add tests for other TimeUnits like :
* DOW
* DOY
* DECADE
* CENTURY

**Comment by asereda-gs** [comment]
Pls add your full name in parenthesis in commit message.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java (hunk 1 above)
Comment (hsyuan) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: I think 17357 is correct. Postgres also gives this answer.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java (hunk 2 above)
Comment (hsyuan) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: will do

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java (hunk 1 above)
Comment (asereda-gs) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: [Oracle](https://docs.oracle.com/cloud/latest/big-data-discovery-cloud/BDDEQ/ceql_functions_date_extract.htm) and [MySql](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_extract) would return `357`

Comment (hsyuan) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Oracle and SQL server does return 357. But MySQL doesn't support extracting millisecond. Venders have difference implementation.

Comment (hsyuan) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Maybe we should log another JIRA for this.

Comment (asereda-gs) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: I see that `extract` function already follows Postgres semantics in existing tests  (`17357` not `357`).

Not sure if there should be different versions of `extract` depending on current dialect ([SqlOperatorTable](https://github.com/apache/calcite/blob/9721283bd0ce46a337f51a3691585cca8003e399/core/src/main/java/org/apache/calcite/sql/SqlOperatorTable.java)). This is obviously out-of-scope for current PR.

**Comment by hsyuan** [comment]
@asereda-gs I have addressed your comment, please take a look. thanks for your review.

**Comment by asereda-gs** [comment]
Thank you!

**Pull request merged**